pkg/operator/ceph/object/controller.go

```diff
@@ -17,215 +17,329 @@ limitations under the License.
+	Kind:       cephObjectStoreKind,
+	APIVersion: fmt.Sprintf("%s/%s", cephv1.CustomResourceGroup, cephv1.Version),
 }
 
-// StartWatch watches for instances of ObjectStore custom resources and acts on them
-func (c *ObjectStoreController) StartWatch(namespace string, stopCh chan struct{}) {
-	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
-		AddFunc:    c.onAdd,
-		UpdateFunc: c.onUpdate,
-		DeleteFunc: c.onDelete,
-	}
+// ReconcileCephObjectStore reconciles a cephObjectStore object
+type ReconcileCephObjectStore struct {
+	client          client.Client
+	scheme          *runtime.Scheme
+	context         *clusterd.Context
+	cephClusterSpec *cephv1.ClusterSpec
+	cephVersion     cephver.CephVersion
+	clusterInfo     *cephconfig.ClusterInfo
+}
 
-	logger.Infof("start watching object store resources in namespace %s", c.namespace)
-	go k8sutil.WatchCR(ObjectStoreResource, c.namespace, resourceHandlerFuncs, c.context.RookClientset.CephV1().RESTClient(), &cephv1.CephObjectStore{}, stopCh)
+// Add creates a new cephObjectStore Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, context *clusterd.Context) error {
+	return add(mgr, newReconciler(mgr, context))
 }
 
-func (c *ObjectStoreController) onAdd(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Creating object store for an external ceph cluster is disabled because no Ceph image is specified")
-		return
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, context *clusterd.Context) reconcile.Reconciler {
+	// Add the cephv1 scheme to the manager scheme so that the controller knows about it
+	mgrScheme := mgr.GetScheme()
+	cephv1.AddToScheme(mgr.GetScheme())
+
+	return &ReconcileCephObjectStore{
+		client:  mgr.GetClient(),
+		scheme:  mgrScheme,
+		context: context,
 	}
+}
 
-	objectStore, err := getObjectStoreObject(obj)
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		logger.Errorf("failed to get objectstore object. %v", err)
-		return
+		return err
 	}
 
-	c.acquireOrchestrationLock()
-	defer c.releaseOrchestrationLock()
-	updateCephObjectStoreStatus(objectStore.GetName(), objectStore.GetNamespace(), k8sutil.ProcessingStatus, c.context)
+	// Watch for changes on the cephObjectStore CRD object
+	err = c.Watch(&source.Kind{Type: &cephv1.CephObjectStore{TypeMeta: controllerTypeMeta}}, &handler.EnqueueRequestForObject{}, opcontroller.WatchUpdatePredicate())
+	if err != nil {
+		return err
+	}
 
-	if c.clusterSpec.External.Enable {
-		_, err := controller.ValidateCephVersionsBetweenLocalAndExternalClusters(c.context, c.namespace, c.clusterInfo.CephVersion)
+	// Watch all other resources
+	for _, t := range objectsToWatch {
```

Comment:
A few questions about the watchers:
- How much filtering of the watches does K8s do compared to what the rook controllers need to filter in the events? If we have to do our own filtering something doesn't seem right.
- With all the new watchers, will the operator need more cpu resources? In v1.1 or 1.2 we already had to double the operator resources. Will we need to increase it again?
- Is it really worth triggering a reconcile for all the child resources? It seems like overkill to me. An admin doesn't just start deleting services and secrets. But if the code is simple/easy to maintain and the resource usage is negligible, it's certainly a good pattern.

Reply:
> A few questions about the watchers:
> 
> * How much filtering of the watches does K8s do compared to what the rook controllers need to filter in the events? If we have to do our own filtering something doesn't seem right.

I've been looking a lot of operators recently and most of them implement their own filtering, so I guess we are just working with the design.

> * With all the new watchers, will the operator need more cpu resources? In v1.1 or 1.2 we already had to double the operator resources. Will we need to increase it again?

Because we are moving to independent controllers, the Rook operator will consume more resources, it's a given but I don't have numbers. And yes, watchers will likely increase CPU consumption. So yes we might need to higher up our recommendation, after some testing.

> * Is it really worth triggering a reconcile for all the child resources? It seems like overkill to me. An admin doesn't just start deleting services and secrets. But if the code is simple/easy to maintain and the resource usage is negligible, it's certainly a good pattern.

I think it is because this brings more robustness to the overall implementation and preserves the desired state no matter what.